File: bnn_engine/tests/test_linear.py
import torch
import sys
from pathlib import Path

# Add the build directory to the path so Python can find the .so file
build_dir = Path(__file__).parent / "build"
sys.path.append(str(build_dir))

# Import your custom SYCL C++ engine!
import bnn_pytorch_ext

def main():
    print("--- BNN SYCL Engine Test ---")
    
    # 1. Define the network shape
    batch_size = 4
    in_features = 256  # Must be a multiple of 64 for this test
    out_features = 128
    
    in_int64s = in_features // 64
    
    # 2. Create dummy data directly in PyTorch
    # We use random integers to simulate packed bit-vectors
    print("Allocating memory...")
    inputs = torch.randint(-1000, 1000, (batch_size, in_int64s), dtype=torch.int64)
    weights = torch.randint(-1000, 1000, (out_features, in_int64s), dtype=torch.int64)
    
    # Thresholds for the Batch Norm fusion (int32)
    thresholds = torch.zeros((out_features,), dtype=torch.int32)
    
    print(f"Inputs Shape:  {inputs.shape} (Int64)")
    print(f"Weights Shape: {weights.shape} (Int64)")

    # 3. Fire the SYCL Kernel
    print("\nLaunching SYCL AVX-512 Kernel...")
    try:
        outputs = bnn_pytorch_ext.linear_forward(inputs, weights, thresholds)
        
        print("Kernel execution successful!")
        print(f"Outputs Shape: {outputs.shape} (Int64)")
        print("\nFirst batch output snippet (Packed bits):")
        print(outputs[0])
        
    except Exception as e:
        print("\nKERNEL CRASHED:")
        print(e)

if __name__ == "__main__":
    main()

File: bnn_engine/tests/test_linear.jl
# test_linear.jl
using Printf

const LIB_PATH = joinpath(@__DIR__, "build", "libbnn_engine.so")

# Device API
function bnn_linear_forward_device(inputs::Matrix{UInt64}, weights::Matrix{UInt64}, thresholds::Vector{Int32})
    in_int64s, batch_size = size(inputs)
    _, out_features = size(weights)
    out_int64s = cld(out_features, 64) 
    outputs = Matrix{UInt64}(undef, out_int64s, batch_size)
    
    @ccall LIB_PATH.c_api_bnn_linear_forward_device(
        inputs::Ptr{UInt64}, weights::Ptr{UInt64}, thresholds::Ptr{Int32}, 
        outputs::Ptr{UInt64}, batch_size::Cint, in_int64s::Cint, out_features::Cint
    )::Cvoid
    
    return outputs
end

# Server API
function bnn_linear_forward_server(inputs::Matrix{UInt64}, weights::Matrix{UInt64}, thresholds::Vector{Int32})
    in_int64s, batch_size = size(inputs)
    _, out_features = size(weights)
    out_int64s = cld(out_features, 64) 
    outputs = Matrix{UInt64}(undef, out_int64s, batch_size)
    
    @ccall LIB_PATH.c_api_bnn_linear_forward_server(
        inputs::Ptr{UInt64}, weights::Ptr{UInt64}, thresholds::Ptr{Int32}, 
        outputs::Ptr{UInt64}, batch_size::Cint, in_int64s::Cint, out_features::Cint
    )::Cvoid
    
    return outputs
end

function main()
    println("--- BNN SYCL Engine Test (Julia) ---")
    
    batch_size = 4
    in_features = 256
    out_features = 128
    
    in_int64s = div(in_features, 64)
    
    println("Allocating memory...")
    # Generate random 64-bit integers natively
    inputs = rand(UInt64, in_int64s, batch_size)
    weights = rand(UInt64, in_int64s, out_features)
    thresholds = zeros(Int32, out_features)
    
    println("Inputs Shape (Julia layout):  $(size(inputs))")
    println("Weights Shape (Julia layout): $(size(weights))")
    
    println("\nLaunching SYCL AVX-512 Kernel via @ccall...")
    
    # Fire the kernel
    outputs = bnn_linear_forward_device(inputs, weights, thresholds)
    println(sum(inputs))
    println(sum(weights))
    println(sum(outputs))
    println(sum(thresholds))
    
    println("Kernel execution successful!")
    println("Outputs Shape (Julia layout): $(size(outputs))")

    println("\nFirst batch output snippet (Packed bits in Hex):")
    # Print the first block of the first batch in hex so we can see the bits
    @printf("0x%016X\n", outputs[1, 1])
    println()

    outputs = bnn_linear_forward_server(inputs, weights, thresholds)
    println(sum(inputs))
    println(sum(weights))
    println(sum(outputs))
    println(sum(thresholds))

    println("Kernel execution successful!")
    println("Outputs Shape (Julia layout): $(size(outputs))")
    
    println("\nFirst batch output snippet (Packed bits in Hex):")
    # Print the first block of the first batch in hex so we can see the bits
    @printf("0x%016X\n", outputs[1, 1])
end

main()
